### ISSUE TITLE
Poor elmenetwise_kernel performance becomes critical on small mini-batch sizes

### ISSUE THREAD

**Issue by pyotr777** [pytorch/pytorch#36638]
I am experiencing a huge slowdown of a CNN training on small mini-batch sizes caused by elementwise_kernels. These kernels are called on optimizer updates: optimizer.zero_grad() and optimizer.step(). I suppose they are responsible for the model weight updates.

I have tested with Pytorch 1.4.0 and with 1.6.0a0+1e22717. 

The plots below show the aggregated time of CUDA kernels of one epoch of the CNN training for a range of mini-batch sizes for the two Pytorch versions. 

![](https://user-images.githubusercontent.com/485462/78657698-21ac6e00-7904-11ea-8e9a-9fe78a5bdff7.png)

![logs_kernels_epochtime](https://user-images.githubusercontent.com/485462/79292944-5d9b9080-7f0d-11ea-8351-bd6973d25db7.png)

Elementwise_kernels have numbers 0, 1, 7... On smaller mini-batch sizes they take about a half of the epoch time. The kernel number 0 takes the largest amount of time among CUDA kernels. It is responsible for optimizer updates. 

There is not much difference between the two Pytorch versions. Please note, that the order and the colours of the kernels on the two plots are different.

Tested on Ubuntu 18.04.3 LTS in a Docker container.
GPU: NVIDIA Quadro P2000
CUDA: 10.1
cuDNN: 7.6.3.30-1
Nvidia driver: 430.64

First reported here: https://github.com/pytorch/pytorch/issues/31975#issuecomment-610302768

**Comment by zasdfgbnm** [comment]
Thanks for reporting! I assume the y-axis is the total time for one epoch?

The shape of the plot looks very much like a `y=1/x` curve. So, one explanation of this behavior I can think of could be: When the problem size is small, GPU is not saturated, so the throughput is proportional to the size of the tensor, which is proportional to batch size. Then the total time for one epoch will be inversely proportional to the throughput, therefore inverse proportional to the batch size. So it is a more or less expected behavior to show a slowdown with shape `1/x` function? I mean, the slowdown is not just caused by elementwise_kernels, but by every kernel.

Your plot does show something: at the large batch size, the elementwise kernel 0 is taking ~25% of the time, and this number increases to 45% when you reduce the batch size.

**Comment by pyotr777** [comment]
@zasdfgbnm You are right about `y=1/x` curve, the elementwise_kernel time is proportional to the number of iterations or `1/mini-batch size`.

What seems unnatural here, is how large is the portion of the epoch time taken by the elementwise_kernel. For reference, I attach a similar graph for the same CNN in Tensorflow on the same machine.

<img width="748" alt="logs_kernels_epochtime" src="https://user-images.githubusercontent.com/485462/79335188-8dbe5000-7f5c-11ea-8e1b-4c3733f9e060.png">

I am not sure if ShuffleInTensor3Simple (number 0) kernel is responsible for the model weight updates here, but even if it is, weight updates take much less time than they do in Pytorch.
